Changes in module 'nf-core/spades'
--- modules/nf-core/spades/main.nf
+++ modules/nf-core/spades/main.nf
@@ -29,7 +29,7 @@
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
     def maxmem = task.memory.toGiga()
-    def illumina_reads = illumina ? ( meta.single_end ? "-s $illumina" : "-1 ${illumina[0]} -2 ${illumina[1]}" ) : ""
+    def illumina_reads = illumina ? ( meta.single_end ? "-s $illumina" : "--pe-1 1 ${illumina[0]} --pe-2 1 ${illumina[1]} --pe-m 1 ${illumina[2]} --pe-s 1 ${illumina[3]}" ) : ""
     def pacbio_reads = pacbio ? "--pacbio $pacbio" : ""
     def nanopore_reads = nanopore ? "--nanopore $nanopore" : ""
     def custom_hmms = hmm ? "--custom-hmms $hmm" : ""

************************************************************
