Changes in module 'nf-core/bbmap/bbduk'
--- modules/nf-core/bbmap/bbduk/main.nf
+++ modules/nf-core/bbmap/bbduk/main.nf
@@ -10,11 +10,13 @@
     input:
     tuple val(meta), path(reads)
     path contaminants
+    val keep_singletons
 
     output:
-    tuple val(meta), path('*.fastq.gz'), emit: reads
-    tuple val(meta), path('*.log')     , emit: log
-    path "versions.yml"                , emit: versions
+    tuple val(meta), path('(${prefix}_[12].fastq.gz,${prefix}.fastq.gz)') , emit: reads
+    tuple val(meta), path('*.log')                                        , emit: log
+    path "versions.yml"                                                   , emit: versions
+    tuple val(meta), path('${prefix}_singletons.fastq.gz'), optional: true, emit: singletons
 
     when:
     task.ext.when == null || task.ext.when
@@ -23,7 +25,8 @@
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
     def raw      = meta.single_end ? "in=${reads[0]}" : "in1=${reads[0]} in2=${reads[1]}"
-    def trimmed  = meta.single_end ? "out=${prefix}.fastq.gz" : "out1=${prefix}_1.fastq.gz out2=${prefix}_2.fastq.gz"
+    def outs = keep_singletons ? "outs=${prefix}_singletons.fastq.gz" : ""
+    def trimmed  = meta.single_end ? "out=${prefix}.fastq.gz" : "out1=${prefix}_1.fastq.gz out2=${prefix}_2.fastq.gz ${outs}"
     def contaminants_fa = contaminants ? "ref=$contaminants" : ''
     """
     maxmem=\$(echo \"$task.memory\"| sed 's/ GB/g/g')

************************************************************
